libs/auth/src/signed_hash_validation.ts

```diff
@@ -1,95 +1,123 @@
+import { existsSync } from 'fs';
 import fs from 'fs/promises';
 import { Readable } from 'stream';
-import { assert } from '@votingworks/basics';
+import { assert, sleep } from '@votingworks/basics';
+import {
+  formatElectionHashes,
+  SignedHashValidationQrCodeValue,
+} from '@votingworks/types';
 
 import {
   constructSignedHashValidationConfig,
   SignedHashValidationConfig,
 } from './config';
 import { signMessage } from './cryptography';
-import { FileKey, TpmKey } from './keys';
+import { runCommand } from './shell';
 import { constructPrefixedMessage } from './signatures';
 
-const SIGNED_HASH_VALIDATION_MESSAGE_PAYLOAD_SEPARATOR = '/';
 const SIGNED_HASH_VALIDATION_QR_CODE_VALUE_SEPARATOR = ';';
+const SIGNED_HASH_VALIDATION_MESSAGE_PAYLOAD_SEPARATOR = '/';
+const SYSTEM_HASH_DISPLAY_LENGTH = 10;
+
+const HASHING_SCRIPT_PATH = '/vx/admin/admin-functions/hash-signature.sh';
+
+async function computeSystemHash(): Promise<string> {
+  /* istanbul ignore next */
+  if (existsSync(HASHING_SCRIPT_PATH)) {
+    const stdout = await runCommand([
+      'sudo',
+      HASHING_SCRIPT_PATH,
+      'noninteractive',
+    ]);
+    return stdout.toString('utf-8');
+  }
+
+  // Mock system hash computation in dev
+  await sleep(2000);
+  return 'UNVERIFIED';
+}
+
+interface ElectionRecord {
+  electionDefinition: { ballotHash: string };
+  electionPackageHash: string;
+}
 
 interface SignedHashValidationMachineState {
+  electionRecord?: ElectionRecord;
   machineId: string;
-  ballotHash?: string;
+  softwareVersion: string;
 }
 
 /**
  * Signed hash validation allows election officials to check that a VxSuite machine hasn't been
  * tampered with.
  */
-export class SignedHashValidation {
-  private readonly machineCertPath: string;
-  private readonly machinePrivateKey: FileKey | TpmKey;
-
-  constructor(
-    // Support specifying a custom config for tests
-    /* istanbul ignore next */
-    input: SignedHashValidationConfig = constructSignedHashValidationConfig()
-  ) {
-    this.machineCertPath = input.machineCertPath;
-    this.machinePrivateKey = input.machinePrivateKey;
-  }
+export async function generateSignedHashValidationQrCodeValue(
+  machineState: SignedHashValidationMachineState,
+  /* istanbul ignore next */
+  config: SignedHashValidationConfig = constructSignedHashValidationConfig()
+): Promise<SignedHashValidationQrCodeValue> {
+  const { electionRecord, machineId, softwareVersion } = machineState;
+  const systemHash = (await computeSystemHash()).slice(
+    0,
+    SYSTEM_HASH_DISPLAY_LENGTH
+  );
+  const combinedElectionHash = electionRecord
+    ? formatElectionHashes(
+        electionRecord.electionDefinition.ballotHash,
+        electionRecord.electionPackageHash
+      )
+    : '';
+  const date = new Date();
 
-  async generateQrCodeValue({
+  const messagePayloadComponents: string[] = [
```

Comment:
we should assert that none of the components contain the separators.

Reply:
ah I see you did that below for the other separator... I think we should do the same here :)

Reply:
✅ Good call https://github.com/votingworks/vxsuite/pull/5302/commits/1542496da162d486c421af4059b4f9df3dfb6436